### PR TITLE
Add action for stale agent cleanup

### PR DESCRIPTION
**Pull request by cconard96**
Related to glpi-project/glpi#12060

### CODE REVIEW

File: setup.php [hunk 1]
```diff
@@ -55,6 +55,8 @@ function plugin_init_uninstall() {
       $UNINSTALL_DIRECT_CONNECTIONS_TYPE  = ['Monitor', 'Peripheral', 'Phone',
                                              'Printer'];
 
+      Conf::$defaults['plugin_uninstall_stale_agents_uninstall'] = 0;
+
```
Comment (cedric-anne) [inline setup.php:59]: Seems that it is never used.

Comment (cconard96) [reply setup.php:59]: It at least suppresses the warning/error message when saving that says one or more config options are unknown.

File: setup.php [hunk 2]
```diff
@@ -81,6 +84,45 @@ function plugin_init_uninstall() {
...
+                     return \PluginUninstallModel::dropdown([
+                         'name' => '_stale_agents_uninstall',
+                         'value' => $stale_agents_uninstall ?? 0,
+                         'entity' => $_SESSION['glpiactive_entity'],
+                         'condition' => [
+                             'types_id' => \PluginUninstallModel::TYPE_MODEL_UNINSTALL
+                         ],
+                         'display' => false,
+                     ]);
+                 },
```
Comment (cedric-anne) [inline setup.php:112]: This function should be transformed into a class method and referenced using a callable syntax like `'PluginUninstallConfig::renderStaleAgentConfigField'`.
`setup.php` files should contains less code as possible.

File: setup.php [hunk 3]
```diff
@@ -81,6 +84,45 @@ function plugin_init_uninstall() {
...
+                 'action_callback' => static function (?Agent $agent, array $config, ?CommonDBTM $item): bool {
+                     if ($item === null) {
+                         return false;
+                     }
+                     $stale_agents_uninstall = Config::getConfigurationValue('plugin:uninstall', 'stale_agents_uninstall');
+                     \PluginUninstallUninstall::uninstall(get_class($item), $stale_agents_uninstall, [
+                         get_class($item) => [$item->fields['id'] => true]
+                     ], '');
```
Comment (cedric-anne) [inline setup.php:120]: Logic that alter the item in `PluginUninstallUninstall::uninstall()` should probably be moved into a new method. Indeed, in cron context, we should not generate any output (progress bar and operation summary).

File: inc/uninstall.class.php [hunk 4]
```diff
@@ -93,221 +93,238 @@ static function processMassiveActionsForOneItemtype(MassiveAction $ma, CommonDBT
...
+         $item = new $type();
+         $item->getFromDB($id);
+
+         self::doOneUninstall($model, $transfer, $item, [
+             'type' => $type,
+             'location' => $location
+         ]);
 
          //Plugin hook after uninstall
          Plugin::doHook("plugin_uninstall_after", $item);
```
Comment (cedric-anne) [inline inc/uninstall.class.php:330]: This hook should probably be done in `self::doOneUninstall()`.

File: setup.php [hunk 5]
```diff
@@ -81,6 +84,35 @@ function plugin_init_uninstall() {
...
+                     }
+                     $stale_agents_uninstall = Config::getConfigurationValue('plugin:uninstall', 'stale_agents_uninstall');
+                     $model = new PluginUninstallModel();
+                     $model->getConfig($stale_agents_uninstall);
+                     $transfer = new Transfer();
+                     $transfer->getFromDB($model->fields["transfers_id"]);
+                     \PluginUninstallUninstall::doOneUninstall($model, $transfer, $item);
+                     return true;
+                 }
+             ]
```
Comment (cedric-anne) [inline setup.php:113]: Same remark here, logic should be moved in corresponding class.

**Comment by AdrienClairembault** [comment]
Is this PR still in draft or is it ready for reviews ?

**Comment by cconard96** [comment]
It is ready for review.

File: inc/config.class.php [hunk 6]
```diff
@@ -143,4 +143,36 @@ static function uninstall() {
 
       return true;
    }
-}
\ No newline at end of file
+
+   public static function preConfigSet(Config $config)
+   {
+       if (isset($config['context']) && $config->input['context'] === 'inventory') {
```
Comment (AdrienClairembault) [inline inc/config.class.php:149]:
```suggestion
       if (isset($config->input['context']) && $config->input['context'] === 'inventory') {
```

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]

**Comment by AdrienClairembault** [comment]
I'm not sure if the issue is from this PR or the core one, but I can only set the value for `Apply uninstall profile` once.
If I try to edit the field later, the value do not change.

**Comment by cconard96** [comment]
> I'm not sure if the issue is from this PR or the core one, but I can only set the value for `Apply uninstall profile` once. If I try to edit the field later, the value do not change.

There is an issue with how the different context for the config value was handled. It is set to 'inventory' when it is passed to `Config::setConfigurationValues` since that is the default for the form. The method doesn't see it in the DB so it does an `add` instead of update. The `add` process is intercepted by the plugin and the context is changed to `plugin:uninstall`. At that point, it conflicts with the record in the database if one was already set and there is a unicity key.

**Review by AdrienClairembault — APPROVED** [review]

**Review by cedric-anne — APPROVED** [review]

**Comment by cedric-anne** [comment]
I merged this in a feature/glpi-10.1 branch as this code is not compatible with GLPI 10.0.

**Pull request merged**